In [0]:
from __future__ import print_function
#import keras
import tensorflow
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
import keras.utils
batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)


Using TensorFlow backend.


11493376/11490434 [==============================] - 1s 0us/step
x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [0]:
# Credits: https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py



model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Instructions for updating:
Use tf.cast instead.
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 13s 222us/step - loss: 0.2610 - acc: 0.9196 - val_loss: 0.0574 - val_acc: 0.9813
Epoch 2/12
60000/60000 [==============================] - 9s 143us/step - loss: 0.0868 - acc: 0.9743 - val_loss: 0.0401 - val_acc: 0.9866
Epoch 3/12
60000/60000 [==============================] - 8s 141us/step - loss: 0.0644 - acc: 0.9808 - val_loss: 0.0382 - val_acc: 0.9874
Epoch 4/12
60000/60000 [==============================] - 8s 141us/step - loss: 0.0507 - acc: 0.9845 - val_loss: 0.0335 - val_acc: 0.9884
Epoch 5/12
60000/60000 [==============================] - 9s 142us/step - loss: 0.0439 - acc: 0.9868 - val_loss: 0.0309 - val_acc: 0.9896
Epoch 6/12
60000/60000 [

## Using tanh activation function with kernel size(4,4)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4),
                 activation='tanh',
                 input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 8s 137us/step - loss: 0.2540 - acc: 0.9228 - val_loss: 0.0600 - val_acc: 0.9810
Epoch 2/12
60000/60000 [==============================] - 8s 129us/step - loss: 0.1065 - acc: 0.9682 - val_loss: 0.0477 - val_acc: 0.9852
Epoch 3/12
60000/60000 [==============================] - 8s 129us/step - loss: 0.0827 - acc: 0.9752 - val_loss: 0.0415 - val_acc: 0.9858
Epoch 4/12
60000/60000 [==============================] - 8s 129us/step - loss: 0.0679 - acc: 0.9798 - val_loss: 0.0432 - val_acc: 0.9876
Epoch 5/12
60000/60000 [==============================] - 8s 128us/step - loss: 0.0596 - acc: 0.9825 - val_loss: 0.0382 - val_acc: 0.9887
Epoch 6/12
60000/60000 [==============================] - 8s 128us/step - loss: 0.0534 - acc: 0.9839 - val_loss: 0.0477 - val_acc: 0.9866
Epoch 7/12
60000/60000 [==============================] - 8s 128us/step - loss: 0.0477 - acc: 0.9856 - val_loss: 0.0392 - 

## Using tanh function with kernel size(5,5)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(5, 5),
                 activation='tanh',
                 input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='tanh'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 8s 131us/step - loss: 0.2603 - acc: 0.9209 - val_loss: 0.0634 - val_acc: 0.9807
Epoch 2/12
60000/60000 [==============================] - 7s 121us/step - loss: 0.1124 - acc: 0.9667 - val_loss: 0.0486 - val_acc: 0.9846
Epoch 3/12
60000/60000 [==============================] - 7s 121us/step - loss: 0.0878 - acc: 0.9736 - val_loss: 0.0442 - val_acc: 0.9856
Epoch 4/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0710 - acc: 0.9793 - val_loss: 0.0429 - val_acc: 0.9877
Epoch 5/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0623 - acc: 0.9808 - val_loss: 0.0316 - val_acc: 0.9897
Epoch 6/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0557 - acc: 0.9827 - val_loss: 0.0369 - val_acc: 0.9890
Epoch 7/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0512 - acc: 0.9840 - val_loss: 0.0370 - 

## Using relu activation function with kernel size(4,4)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4, 4),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 8s 139us/step - loss: 0.2324 - acc: 0.9276 - val_loss: 0.0481 - val_acc: 0.9844
Epoch 2/12
60000/60000 [==============================] - 8s 130us/step - loss: 0.0784 - acc: 0.9766 - val_loss: 0.0378 - val_acc: 0.9877
Epoch 3/12
60000/60000 [==============================] - 8s 130us/step - loss: 0.0584 - acc: 0.9828 - val_loss: 0.0302 - val_acc: 0.9896
Epoch 4/12
60000/60000 [==============================] - 8s 132us/step - loss: 0.0477 - acc: 0.9863 - val_loss: 0.0301 - val_acc: 0.9899
Epoch 5/12
60000/60000 [==============================] - 8s 131us/step - loss: 0.0410 - acc: 0.9873 - val_loss: 0.0233 - val_acc: 0.9920
Epoch 6/12
60000/60000 [==============================] - 8s 130us/step - loss: 0.0374 - acc: 0.9888 - val_loss: 0.0237 - val_acc: 0.9920
Epoch 7/12
60000/60000 [==============================] - 8s 130us/step - loss: 0.0326 - acc: 0.9900 - val_loss: 0.0234 - 

## Using relu activation function with kernel size(5,5)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(5, 5),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 8s 132us/step - loss: 0.1924 - acc: 0.9405 - val_loss: 0.0476 - val_acc: 0.9837
Epoch 2/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0576 - acc: 0.9826 - val_loss: 0.0395 - val_acc: 0.9866
Epoch 3/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0401 - acc: 0.9876 - val_loss: 0.0285 - val_acc: 0.9904
Epoch 4/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0323 - acc: 0.9901 - val_loss: 0.0290 - val_acc: 0.9897
Epoch 5/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0258 - acc: 0.9923 - val_loss: 0.0215 - val_acc: 0.9922
Epoch 6/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0223 - acc: 0.9928 - val_loss: 0.0268 - val_acc: 0.9905
Epoch 7/12
60000/60000 [==============================] - 7s 122us/step - loss: 0.0184 - acc: 0.9940 - val_loss: 0.0229 - 

### With 3 convolution layer:

##### with kernel size (3,3)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Dropout(0.5))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s 205us/step - loss: 0.2105 - acc: 0.9341 - val_loss: 0.0557 - val_acc: 0.9822
Epoch 2/12
60000/60000 [==============================] - 12s 192us/step - loss: 0.0791 - acc: 0.9760 - val_loss: 0.0362 - val_acc: 0.9885
Epoch 3/12
60000/60000 [==============================] - 12s 193us/step - loss: 0.0584 - acc: 0.9823 - val_loss: 0.0405 - val_acc: 0.9868
Epoch 4/12
60000/60000 [==============================] - 11s 191us/step - loss: 0.0484 - acc: 0.9848 - val_loss: 0.0305 - val_acc: 0.9898
Epoch 5/12
60000/60000 [==============================] - 11s 192us/step - loss: 0.0422 - acc: 0.9869 - val_loss: 0.0257 - val_acc: 0.9918
Epoch 6/12
60000/60000 [==============================] - 12s 192us/step - loss: 0.0380 - acc: 0.9886 - val_loss: 0.0274 - val_acc: 0.9916
Epoch 7/12
60000/60000 [==============================] - 11s 191us/step - loss: 0.0346 - acc: 0.9888 - val_loss: 0.

##### with kernel size (4,4)

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4,4),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Dropout(0.5))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s 200us/step - loss: 0.2115 - acc: 0.9342 - val_loss: 0.0463 - val_acc: 0.9852
Epoch 2/12
60000/60000 [==============================] - 11s 185us/step - loss: 0.0858 - acc: 0.9736 - val_loss: 0.0366 - val_acc: 0.9867
Epoch 3/12
60000/60000 [==============================] - 11s 185us/step - loss: 0.0637 - acc: 0.9803 - val_loss: 0.0350 - val_acc: 0.9890
Epoch 4/12
60000/60000 [==============================] - 11s 186us/step - loss: 0.0560 - acc: 0.9832 - val_loss: 0.0345 - val_acc: 0.9887
Epoch 5/12
60000/60000 [==============================] - 11s 186us/step - loss: 0.0456 - acc: 0.9855 - val_loss: 0.0326 - val_acc: 0.9890
Epoch 6/12
60000/60000 [==============================] - 11s 185us/step - loss: 0.0400 - acc: 0.9874 - val_loss: 0.0288 - val_acc: 0.9907
Epoch 7/12
60000/60000 [==============================] - 11s 186us/step - loss: 0.0373 - acc: 0.9887 - val_loss: 0.

### with single layer of dropout

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4,4),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='relu'))
model.add(Conv2D(64, (4, 4), activation='relu'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Flatten())
model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s 197us/step - loss: 0.1820 - acc: 0.9434 - val_loss: 0.0368 - val_acc: 0.9882
Epoch 2/12
60000/60000 [==============================] - 11s 181us/step - loss: 0.0492 - acc: 0.9849 - val_loss: 0.0313 - val_acc: 0.9897
Epoch 3/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0352 - acc: 0.9892 - val_loss: 0.0252 - val_acc: 0.9919
Epoch 4/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0263 - acc: 0.9917 - val_loss: 0.0279 - val_acc: 0.9908
Epoch 5/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0227 - acc: 0.9927 - val_loss: 0.0270 - val_acc: 0.9911
Epoch 6/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0180 - acc: 0.9942 - val_loss: 0.0278 - val_acc: 0.9920
Epoch 7/12
60000/60000 [==============================] - 11s 181us/step - loss: 0.0151 - acc: 0.9956 - val_loss: 0.

##### without dropout layer

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4,4),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (4, 4), activation='relu'))
model.add(Conv2D(64, (4, 4), activation='tanh'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Flatten())
model.add(Dense(128, activation = 'tanh'))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s 196us/step - loss: 0.1312 - acc: 0.9587 - val_loss: 0.0397 - val_acc: 0.9887
Epoch 2/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0363 - acc: 0.9893 - val_loss: 0.0367 - val_acc: 0.9880
Epoch 3/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0217 - acc: 0.9937 - val_loss: 0.0265 - val_acc: 0.9912
Epoch 4/12
60000/60000 [==============================] - 11s 181us/step - loss: 0.0124 - acc: 0.9970 - val_loss: 0.0239 - val_acc: 0.9926
Epoch 5/12
60000/60000 [==============================] - 11s 181us/step - loss: 0.0076 - acc: 0.9983 - val_loss: 0.0248 - val_acc: 0.9922
Epoch 6/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0043 - acc: 0.9991 - val_loss: 0.0214 - val_acc: 0.9935
Epoch 7/12
60000/60000 [==============================] - 11s 181us/step - loss: 0.0022 - acc: 0.9997 - val_loss: 0.

##### with (5,5) size kernel

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(4,4),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (5, 5), activation='relu'))
model.add(Conv2D(64, (5, 5), activation='tanh'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Flatten())
model.add(Dense(128, activation = 'tanh'))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s 195us/step - loss: 0.1369 - acc: 0.9584 - val_loss: 0.0392 - val_acc: 0.9875
Epoch 2/12
60000/60000 [==============================] - 10s 175us/step - loss: 0.0388 - acc: 0.9883 - val_loss: 0.0349 - val_acc: 0.9879
Epoch 3/12
60000/60000 [==============================] - 10s 175us/step - loss: 0.0234 - acc: 0.9931 - val_loss: 0.0308 - val_acc: 0.9888
Epoch 4/12
60000/60000 [==============================] - 10s 175us/step - loss: 0.0147 - acc: 0.9959 - val_loss: 0.0248 - val_acc: 0.9919
Epoch 5/12
60000/60000 [==============================] - 10s 175us/step - loss: 0.0086 - acc: 0.9980 - val_loss: 0.0245 - val_acc: 0.9917
Epoch 6/12
60000/60000 [==============================] - 11s 176us/step - loss: 0.0056 - acc: 0.9990 - val_loss: 0.0225 - val_acc: 0.9927
Epoch 7/12
60000/60000 [==============================] - 11s 176us/step - loss: 0.0034 - acc: 0.9993 - val_loss: 0.

## with 5 convolution layer

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(3, 3)))
model.add(Dropout(0.5))

model.add(Flatten())

model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 23s 383us/step - loss: 0.2696 - acc: 0.9131 - val_loss: 0.0502 - val_acc: 0.9848
Epoch 2/12
60000/60000 [==============================] - 21s 356us/step - loss: 0.0817 - acc: 0.9752 - val_loss: 0.0315 - val_acc: 0.9892
Epoch 3/12
60000/60000 [==============================] - 21s 354us/step - loss: 0.0609 - acc: 0.9812 - val_loss: 0.0333 - val_acc: 0.9891
Epoch 4/12
60000/60000 [==============================] - 21s 355us/step - loss: 0.0525 - acc: 0.9840 - val_loss: 0.0308 - val_acc: 0.9890
Epoch 5/12
60000/60000 [==============================] - 21s 357us/step - loss: 0.0415 - acc: 0.9875 - val_loss: 0.0243 - val_acc: 0.9909
Epoch 6/12
60000/60000 [==============================] - 21s 356us/step - loss: 0.0390 - acc: 0.9877 - val_loss: 0.0209 - val_acc: 0.9933
Epoch 7/12
60000/60000 [==============================] - 21s 355us/step - loss: 0.0357 - acc: 0.9892 - val_loss: 0.

#### changing the activation function in various layers 

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.5))

model.add(Flatten())

model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 23s 388us/step - loss: 0.2338 - acc: 0.9295 - val_loss: 0.0803 - val_acc: 0.9765
Epoch 2/12
60000/60000 [==============================] - 22s 365us/step - loss: 0.0766 - acc: 0.9768 - val_loss: 0.1230 - val_acc: 0.9651
Epoch 3/12
60000/60000 [==============================] - 22s 367us/step - loss: 0.0587 - acc: 0.9826 - val_loss: 0.0277 - val_acc: 0.9914
Epoch 4/12
60000/60000 [==============================] - 22s 366us/step - loss: 0.0472 - acc: 0.9856 - val_loss: 0.0471 - val_acc: 0.9853
Epoch 5/12
60000/60000 [==============================] - 22s 365us/step - loss: 0.0425 - acc: 0.9871 - val_loss: 0.0216 - val_acc: 0.9932
Epoch 6/12
60000/60000 [==============================] - 22s 364us/step - loss: 0.0381 - acc: 0.9887 - val_loss: 0.0201 - val_acc: 0.9928
Epoch 7/12
60000/60000 [==============================] - 22s 364us/step - loss: 0.0345 - acc: 0.9898 - val_loss: 0.

### changing the architecture

In [0]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='tanh',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='tanh'))
model.add(Conv2D(96, (3, 3), activation='tanh'))
model.add(Conv2D(128, (3, 3), activation='relu'))
model.add(Conv2D(96, (3, 3), activation='tanh'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())

model.add(Dense(128, activation = 'relu'))
model.add(Dropout(0.25))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])

model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 34s 569us/step - loss: 0.2166 - acc: 0.9332 - val_loss: 0.0584 - val_acc: 0.9814
Epoch 2/12
60000/60000 [==============================] - 32s 526us/step - loss: 0.0657 - acc: 0.9803 - val_loss: 0.0340 - val_acc: 0.9896
Epoch 3/12
60000/60000 [==============================] - 32s 528us/step - loss: 0.0473 - acc: 0.9862 - val_loss: 0.0291 - val_acc: 0.9894
Epoch 4/12
60000/60000 [==============================] - 32s 529us/step - loss: 0.0389 - acc: 0.9879 - val_loss: 0.0264 - val_acc: 0.9920
Epoch 5/12
60000/60000 [==============================] - 32s 528us/step - loss: 0.0332 - acc: 0.9897 - val_loss: 0.0297 - val_acc: 0.9914
Epoch 6/12
60000/60000 [==============================] - 32s 528us/step - loss: 0.0280 - acc: 0.9914 - val_loss: 0.0333 - val_acc: 0.9899
Epoch 7/12
60000/60000 [==============================] - 32s 527us/step - loss: 0.0251 - acc: 0.9926 - val_loss: 0.

# Conclusion :

## For 2 CNN Layers :
1. tanh activation function with kernel size (4,4) (1) Test loss: 0.04084764857832672 (2) Test accuracy: 0.9895
2. tanh function with kernel size(5,5)  (1) Test loss: 0.029969597521997456  (2) Test accuracy: 0.9924
3. relu activation function with kernel size(4,4) (1) Test loss: 0.023073950349439928 (2) Test accuracy: 0.9925
4. relu activation function with kernel size(5,5) (1) Test loss: 0.024874476279205828 (2) Test accuracy: 0.993

## For 3 CNN Layers : 
1. tanh activation function with kernel size (3,3)  (1) Test loss: 0.024087643385360024 (2) Test accuracy: 0.9931
2. tanh activation function with kernel size (3,3)   (1) Test loss: 0.028245600645799278 (2) Test accuracy: 0.9912

#### With signle layer of dropout
1. relu activation function with kernel size(4,4)  (1) Test loss: 0.028049952627918127 (2) Test accuracy: 0.9933

#### Without dropout layer
1. tanh, relu and tanh activation function with kernel size(4,4) (1) Test loss: 0.02414796507277315 (2) Test accuracy: 0.9934 
2. tanh, relu and tanh activation function with kernel size(4,4) and (5,5), (5,5) (1) Test loss: 0.023093068777823283 (2) Test accuracy: 0.9936

## 5 CNN Layers :
1. tanh and relu activation function with k size (3,3) (1) Test loss: 0.02295284935171585 (2) Test accuracy: 0.9931
2. Changing the activation function in diff layers (1) Test loss: 0.020912341866253518 (2) Test accuracy: 0.9932
3. changing the number of neurons : (1) Test loss: 0.02878502955801289 (2) Test accuracy: 0.9934

Steps Taken :
  1. Get the dataset.
  2. Define the model architecture.
  3. Compile the model.
  4. Calculate the accuracy of the model.
  5. Here I have tried various architecture with diff-diff activation function, different convolution layer, and different drop out rate.